pyanaconda/core/subscription.py

```diff
@@ -38,3 +53,191 @@ def check_system_purpose_set(sysroot="/"):
     """
     syspurpose_path = util.join_paths(sysroot, RHSM_SYSPURPOSE_FILE_PATH)
     return os.path.exists(syspurpose_path)
+
+# Asynchronous registration + subscription & unregistration handling
+#
+# The Red Hat subscription related tasks communicate over network and might
+# take some time to finish (up to tens of seconds). We definitely don't want
+# to block either automated installation or the UI before they finish.
+#
+# These tasks (register + attach subscription and unregister) need to run in
+# threads and these threads need to be started from at least two places:
+# - from early startup code for automated installations
+# - from Subscription spoke based on user interaction
+#
+# Also in some cases, multiple individual DBus tasks will need to be run
+# in sequence with any errors handled accordingly.
+#
+# Anaconda modularity is not yet advanced enough to handle this in a generic
+# manner, so we need simple scheduler living in the context of the main Anaconda
+# thread. The simple scheduler hosts the code that starts the respective subscription
+# handling thread, which assures appropriate tasks are run.
+#
+# As the scheduler code than can be run either during early startup or in reaction to user
+# interaction in the Subscription spoke we avoid code duplication.
+
+
+def dummy_progress_callback(phase):
+    """Dummy progress reporting function used if no custom callback is set."""
+    pass
+
+
+def dummy_error_callback(error_message):
+    """Dummy error reporting function used if no custom callback is set."""
+    pass
+
+
+def register_and_subscribe(progress_callback=None, error_callback=None):
```

Comment:
I see this a too big method to be just one method. If I'm counting correctly it can be split to about 5 parts.

I won't block this PR on this.

Reply:
This is basically a recipe on how to register and subscribe a system, that should consist of individual simple steps. Splitting it into individual private micro methods would make it harder to follow and introduce a lot of boilerplate.

Reply:
Also, longer term we will likely want to use some generic way of running a batch of DBus tasks both during installation and cases like this with runtime tasks.